### ISSUE TITLE
`useEvent` cannot be used like a subscriber

### ISSUE THREAD

**Issue by JordanMartinez** [JordanMartinez/purescript-halogen-hooks-extra#2]
`useEvent` works when one needs to define a callback that should run every time a new value is pushed to that callback. However, that callback cannot function as a subscriber. 

By subscriber, I mean the handler can set up some resources that are used to handle events and later clean up those resources when it no longer needs to handle events.

The current implementation of `useEvent` can be used to handle events. However, any resources one sets up in the callback cannot be easily cleaned up before setting a new callback.

**Comment by JordanMartinez** [comment]
Here's the only workaround I can think of using `Ref`s when using lifecyle intializers and finalizers:
```purescript
component :: H.Component HH.HTML (Const Void) Unit Unit Aff
component = Hooks.component \_ -> Hooks.do
  _ /\ unsubscribeRef <- useRef (Nothing :: Maybe (HookM _ _ _ Unit))
  changes <- useEvent

  let triggerCallback = changes.push
  -- at some point in code, we trigger the callback via `triggerCallback`

  useLifecycleEffect do
    changes.setCallback $ Just \i -> do
      -- code for event handler
      liftEffect $ log $ "New value is: " <> show i

      -- code for subscription stuff
      mbCallback <- liftEffect $ Ref.read unsubscribeRef
      case mbCallback of
        Nothing -> do
          -- set up some resources for the first time

          -- and then store the unsubscriber in a Ref
          let unsubscribe = pure unit
          liftEffect $ Ref.write unsubscribe unsubscribeRef
        Just _ -> do
          -- we already set up resources, so don't do it again
          pure unit

    pure $ Just do
      mbCallback <- liftEffect $ Ref.read unsubscribeRef
      case mbCallback of
        Just unsubscribe -> do
          -- haven't yet unsubscribed, so we need to do it
          unsubscribe
        Nothing -> do
          -- already unsubscribed, so no need to do it again
          pure unit

  Hooks.pure $ HH.text "example"
```

**Comment by JordanMartinez** [comment]
And here's the code for `useTickEffect` following this same idea. I believe this should work:
```purescript
component :: H.Component HH.HTML (Const Void) Unit Unit Aff
component = Hooks.component \_ -> Hooks.do
  _ /\ unsubscribeRef <- useRef (Nothing :: Maybe (HookM _ _ _ Unit))
  changes <- useEvent
  state /\ tState <- useState 0

  let triggerCallback = changes.push
  -- at some point in code, we trigger the callback via `triggerCallback`

  Hooks.captures { state } useTickEffect do
    -- (if needed) unsubscribe to cleanup resources
    -- and be sure to do this first BEFORE setting the callback.
    mbCallback <- liftEffect $ Ref.read unsubscribeRef
    case mbCallback of
      Just unsubscribe -> do
        -- haven't yet unsubscribed, so we need to do it
        unsubscribe
      Nothing -> do
        -- already unsubscribed, so no need to do it again
        -- OR never subscribed, so nothing to do...
        pure unit

    changes.setCallback $ Just \i -> do
      -- code for event handler
      liftEffect $ log $ "New value is: " <> show i

      -- code for subscription stuff
      mbCallback <- liftEffect $ Ref.read unsubscribeRef
      case mbCallback of
        Nothing -> do
          -- set up some resources for the first time

          -- and then store the unsubscriber in a Ref
          let unsubscribe = pure unit
          liftEffect $ Ref.write unsubscribe unsubscribeRef
        Just _ -> do
          -- we already set up resources, so don't do it again
          pure unit

    pure Nothing

  Hooks.pure $ HH.text "example"
```

**Comment by JordanMartinez** [comment]
Fixed by #3
